File: ios/WatchAppBackend Watch App/Model/Data/Statistics/WatchData/Motion/Attitude.swift
//
//  Attitude.swift
//  WatchApp Watch App
//
//  Created by Mikhail Plekunov on 11/15/23.
//

import Foundation

struct Attitude : Codable, Equatable {
    var roll: Measurement<UnitAngle>
    var yaw: Measurement<UnitAngle>
    var pitch: Measurement<UnitAngle>
    
    init(roll: Measurement<UnitAngle>, yaw: Measurement<UnitAngle>, pitch: Measurement<UnitAngle>) {
        self.roll = roll
        self.yaw = yaw
        self.pitch = pitch
    }
    
    enum CodingKeys: CodingKey {
        case roll
        case yaw
        case pitch
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(self.roll, forKey: .roll)
        try container.encode(self.yaw, forKey: .yaw)
        try container.encode(self.pitch, forKey: .pitch)
    }
    
    static func == (lhs: Self, rhs: Self) -> Bool {
        return lhs.pitch == rhs.pitch &&
        lhs.roll == rhs.roll &&
        lhs.yaw == rhs.yaw
    }
}

File: backend/model/WaterSkiing/Course/attitude.ts
import { Measurement } from "../../Units/unit";
import { UnitAngle } from "../../Units/unitAngle";

export class Attitude {
    roll: Measurement<UnitAngle>
    yaw: Measurement<UnitAngle>
    pitch: Measurement<UnitAngle>

    constructor(
        roll: Measurement<UnitAngle>,
        yaw: Measurement<UnitAngle>,
        pitch: Measurement<UnitAngle>
    ) {
        this.pitch = pitch
        this.roll = roll
        this.yaw = yaw
    }

    static parse(json: string): Attitude {
        try {
            let anyJSON = JSON.parse(json)

            let roll = new Measurement(anyJSON.roll.value as number, UnitAngle.parse(JSON.stringify(anyJSON.roll.unit)))
            let yaw = new Measurement(anyJSON.yaw.value as number, UnitAngle.parse(JSON.stringify(anyJSON.yaw.unit)))
            let pitch = new Measurement(anyJSON.pitch.value as number, UnitAngle.parse(JSON.stringify(anyJSON.pitch.unit)))
        
            return new Attitude(roll, yaw, pitch)
        } catch (error) {
            throw new Error(`Attitude - ${error}`)
        }
    }
}